sycl/doc/design/ESIMDDesignNotes.md

```diff
@@ -0,0 +1,277 @@
+
+Source locations:
+
+- `clang/lib/Sema/SemaSYCL.cpp`
+- `clang/lib/Sema/SemaDecl.cpp`
+- `clang/lib/Sema/SemaExpr.cpp`
+- `clang/include/clang/Basic/Attr.td`
+- `clang/include/clang/Basic/DiagnosticSemaKinds.td`
+
+#### Code (LLVMIR) generator
+
+ESIMD-specific code generator tweaks are mostly translations of internal FE
+representation of variaous ESIMD attributes into LLVM IR attributes or metadata.
+
+##### Kernel signature generation
+
+For ESIMD kernels, a number of additional attributes are generated for the
+kernel function itself as well as certain argument.
+
+- Kernels are annotated with `sycl_explicit_simd` and
+  `intel_reqd_sub_group_size` attributes. The latter must always be `1` for a
+  ESIMD kernel or function.
+- An argument which conveys accessor's pointer is assigned a
+  `kernel_arg_accessor_ptr` attribute
+
+##### Global variable code generation
+
+ESIMD supports "private globals" - global variables which have one copy per
+thread of execution (similar to C++ thread_local), normally allocated of Gen
+register file. To make a global variable a "private global",
+`__attribute__((opencl_private)) __attribute__((sycl_explicit_simd))`
+attributes are used. Globals of this can be forced to a specific register using
+the `__attribute__((register_num(n)))` attribute. The clang code generator
+translates these to `genx_volatile` and `genx_byte_offset` LLVM IR attributes.
+
+##### Function attributes translations
+
+- `sycl_esimd_vectorize` -> `CMGenxSIMT`
+
+Source locations:
+
+- `clang/lib/CodeGen/CGSYCLRuntime.cpp`
+- `clang/lib/CodeGen/CodeGenFunction.cpp`
+- `clang/lib/CodeGen/CodeGenModule.cpp`
+
+### Clang middle-end
+
+#### ESIMD API restriction verifier
+
+This component is an LLVM IR pass over a compiled translation unit. It checks
+for presence of certain SYCL APIs which are disallowed within ESIMD code. For
+exaple, SYCL reductions are not allowed in ESIMD. The verifier does this by
+demangling all the call targets within ESIMD code and matching them with
+internal sub-string filters. Invoked from `clang/lib/CodeGen/BackendUtil.cpp`.
+
+Source locations:
+
+- `llvm/lib/SYCLLowerIR/ESIMD/ESIMDVerifier.cpp`
+
+### SYCL post-link tool
+
+As a part of the input device code module transformation pipeline, the
+post-link tool splits the input module (or modules resulting from splitting
+by other characteristics, such as aspects) into two - SYCL and ESIMD ones.
+Shared functions invoked both from SYCL and ESIMD are cloned during the process.
+This is necessary because SYCL and ESIMD parts must undergo different set of
+transformations before generating resulting SPIR-V. ESIMD modules resulting from
+splitting are marked with specific device binary property `isEsimdImage` (see
+[source](https://github.com/intel/llvm/blob/9dc14a292f479880b5ab2e95f99a8414b31d1165/llvm/tools/sycl-post-link/sycl-post-link.cpp#L453)
+.)
+
+#### Driver
```

Comment:
This section should be renamed. We already have another component with the name driver and any other document references `sycl-post-link` as a driver. Probably we can merge this sub-section the parent section above.
@v-klochkov, what do you think?

Reply:
Done in https://github.com/intel/llvm/pull/7824/commits/1b3a19f059878edb03ce1314b28dbd4d9dfc1871. Please, let me know if it looks good to you.

Reply:
Looks good to me. Thank you for doing these additional changes!